include/zephyr/drivers/mspi.h

```diff
@@ -0,0 +1,791 @@
+/*
+ * Copyright (c) 2024, Ambiq Micro Inc. <www.ambiq.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public APIs for MSPI driver
+ */
+
+#ifndef ZEPHYR_INCLUDE_MSPI_H_
+#define ZEPHYR_INCLUDE_MSPI_H_
+
+#include <errno.h>
+
+#include <zephyr/sys/__assert.h>
+#include <zephyr/types.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief MSPI Driver APIs
+ * @defgroup mspi_interface MSPI Driver APIs
+ * @ingroup io_interfaces
+ * @{
+ */
+
+/**
+ * @brief MSPI operational mode
+ */
+enum mspi_op_mode {
+	MSPI_OP_MODE_MASTER         = 0,
+	MSPI_OP_MODE_SLAVE          = 1,
+};
+
+/**
+ * @brief MSPI duplex mode
+ */
+enum mspi_duplex {
+	MSPI_HALF_DUPLEX            = 0,
+	MSPI_FULL_DUPLEX            = 1,
+};
+
+/**
+ * @brief MSPI I/O mode capabilities
+ */
+enum mspi_io_mode {
```

Comment:
Mentioned this below as well, but I don't think setting I/O mode this way makes sense. There are going to be cases where some commands for the device can run in 1-4-4 mode, and other commands are limited to 1-1-4 mode (or 1-1-1) mode for example. I think we need to set the *number of pads* each time we issue a SPI transfer

Reply:
Please see my explanation in https://github.com/zephyrproject-rtos/zephyr/pull/71769#discussion_r1596177507. 
When the mode of operation changes, it may be required to also reconfigure the device by sending commands, so calling `mspi_dev_config` before hand should be most appropriate.